### PR TITLE
Tokenizer and get_lm refactoring

### PR DESCRIPTION
**Pull request by yzhang123**
- [x]  renamed NeMoBERT tokenizer to be AutoTokenizer, instead of instantiating only a small subset of specific tokenizers now supports Huggingface AutoTokenizer which has a larger variety

- [x] call tokenizers are compatible with BERT-like models. this includes recently added "WordTokenizer", "CharTokenizer". Added ci tests.

- [x] create_spc_tokenizer():
 Sentence piece tokenizer creation can be now configured based on algorithm "bpe", "unigram", "word", "char" following Google sentence piece tokenizer Api

- [x] refactored and fixed problems in helper functions get_tokenizer(), get_lm_model() - high-level user API - to be more user friendly

- [x] get_tokenizer():
     input parameters:
     tokenizer_name
     tokenizer_model,
     special_tokens,
     vocab_file,
     pretrained_model_name, -> removed
     vocab_size->removed,
     do_lower_case->removed,
     data_file-> removed
      sample_size->removed,

### CODE REVIEW

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (okuchaiev) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:62]: Please search for this classname in docstrings

Comment (yzhang123) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py]: done thanks!

**Review by okuchaiev — APPROVED** [review]
Looks good to me. Thanks!

File: examples/nlp/glue_benchmark/glue_benchmark_config.yaml
Comment (ekmb) [inline examples/nlp/glue_benchmark/glue_benchmark_config.yaml:45]: or sentencepiece or word or char - or refer to a method or a way to get the list of supported tokenizers

File: examples/nlp/glue_benchmark/glue_benchmark_config.yaml
Comment (ekmb) [inline examples/nlp/glue_benchmark/glue_benchmark_config.yaml:52]: let's discuss in the upcoming meeging if we can reduce/remove/regroup these sentencepiece only related args

File: nemo/collections/asr/models/ctc_bpe_models.py
Comment (ekmb) [inline nemo/collections/asr/models/ctc_bpe_models.py:115]: @titu1994 why you don't like this solution? this is WPE

File: nemo/collections/common/tokenizers/char_tokenizer.py
Comment (ekmb) [inline nemo/collections/common/tokenizers/char_tokenizer.py:107]: please fix the add_special_tokens() method to return the number of special tokens added so that we can later adjust the model size properly. Please see nemoGPT2 implementation https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/common/tokenizers/gpt2_tokenizer.py#L57 and HF:https://github.com/huggingface/transformers/blob/77abd1e79fc37efb45bde0041f9dad0eabc55517/src/transformers/tokenization_utils_base.py#L673 and

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (ekmb) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:45]: are we using this anywhere?

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (ekmb) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:63]: are we using this anywhere?

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (ekmb) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:111]: We used do_lower_case for SPE only, in the old setup and it was ignored if it didn't match the pretrained_model_name. Do we need it here?

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py
Comment (ekmb) [inline nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:127]: why are you removing this?

File: nemo/collections/common/tokenizers/tokenizer_spec.py
Comment (ekmb) [inline nemo/collections/common/tokenizers/tokenizer_spec.py:57]: please see the previous implementation in the Sentence piece - tokens could be either str or List - I'd leave this change for a separate PR

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (ekmb) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:72]: or we can set a WPE default here @yzhang123 , @titu1994 ?

File: nemo/collections/nlp/models/glue_benchmark/glue_benchmark_model.py
Comment (ekmb) [inline nemo/collections/nlp/models/glue_benchmark/glue_benchmark_model.py:114]: why this can't be a part of get_lm_model?

File: nemo/collections/nlp/models/glue_benchmark/glue_benchmark_model.py
Comment (ekmb) [inline nemo/collections/nlp/models/glue_benchmark/glue_benchmark_model.py:236]: should we define a NLP model and define all such general method there? like an ASR/ctc model as a base class?

File: nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py
Comment (ekmb) [inline nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py:116]: so this will include models user will have to initialize directly by replacing get_lm_model()?

File: nemo/collections/nlp/modules/common/lm_utils.py
Comment (ekmb) [inline nemo/collections/nlp/modules/common/lm_utils.py:45]: can we have the complete list of megatron names back here?

**Review by ekmb — CHANGES_REQUESTED** [review]

File: nemo/collections/nlp/modules/common/megatron/megatron_utils.py
Comment (ekmb) [inline nemo/collections/nlp/modules/common/megatron/megatron_utils.py:187]: typing + docstring

File: examples/nlp/glue_benchmark/glue_benchmark_config.yaml
Comment (yzhang123) [reply examples/nlp/glue_benchmark/glue_benchmark_config.yaml:45]: there is get_tokenizer_list

File: examples/nlp/glue_benchmark/glue_benchmark_config.yaml
Comment (yzhang123) [reply examples/nlp/glue_benchmark/glue_benchmark_config.yaml:52]: done, agreed on removing vocab_size, sample_size, data_file if everything works as expected

File: nemo/collections/common/tokenizers/char_tokenizer.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/char_tokenizer.py:107]: Please sync with @AlexGrinch about that after PR merge

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:45]: tokens_to_text()

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:111]: yes as discussed in meeting

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:127]: its refactored to super class

File: nemo/collections/common/tokenizers/tokenizer_spec.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/tokenizer_spec.py:57]: yes not for this PR

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:72]: we can set "bert-base-cased" as default

File: nemo/collections/nlp/models/glue_benchmark/glue_benchmark_model.py
Comment (yzhang123) [reply nemo/collections/nlp/models/glue_benchmark/glue_benchmark_model.py:236]: yes, in separate PR

File: nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py:116]: we do not imply this function is connected to get_lm_model, only what the function says: all huggingface lm models.
So it includes all huggingface models user can instantiate using get_lm_model

File: nemo/collections/nlp/modules/common/lm_utils.py
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/lm_utils.py:45]: since model_type will be removed, this function will be gone too

File: examples/nlp/glue_benchmark/glue_benchmark_config.yaml
Comment (ekmb) [reply examples/nlp/glue_benchmark/glue_benchmark_config.yaml:45]: could we maybe add this to the comment instead of explicit listing?

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (ekmb) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:72]: if @titu1994 doesn't need a default, I'm fine with not having it

File: nemo/collections/nlp/modules/common/lm_utils.py
Comment (ekmb) [reply nemo/collections/nlp/modules/common/lm_utils.py:45]: wait, why is this function will be gone? we need a way to see the list of available models

File: nemo/collections/common/tokenizers/tokenizer_spec.py
Comment (ekmb) [reply nemo/collections/common/tokenizers/tokenizer_spec.py:57]: then please revert the removal from SPE

File: nemo/collections/nlp/modules/common/lm_utils.py
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/lm_utils.py]: there is another function get_pretrained_lm_models_list

File: nemo/collections/nlp/models/glue_benchmark/glue_benchmark_model.py
Comment (yzhang123) [reply nemo/collections/nlp/models/glue_benchmark/glue_benchmark_model.py]: will remove entire part since tokenizer vocab_size will be removed

File: nemo/collections/common/tokenizers/tokenizer_spec.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/tokenizer_spec.py:57]: why? we are not using it anywhere. we have both the function tokens_to_ids and token_to_id.
Its really bad software design to let token_to_id take a list instead of a str, if you want a list use tokens_to_ids

File: examples/nlp/glue_benchmark/glue_benchmark_config.yaml
Comment (yzhang123) [reply examples/nlp/glue_benchmark/glue_benchmark_config.yaml:45]: why? tokenizer_nameis a required input. This config is supposed to work out of the box, so it makes sense to use model.language_model.pretrained_model_name. which is BERT

**Review by titu1994 — COMMENTED** [review]
Minor comments here and there, but good to go. I'll let @ekmb take final call.

File: examples/nlp/glue_benchmark/glue_benchmark_config.yaml
Comment (yzhang123) [reply examples/nlp/glue_benchmark/glue_benchmark_config.yaml:45]: @titu1994  yes done in commit

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:111]: @titu1994 it depends some huggingface tokenizers do not have do_lower_case member. So yes i will remove do_lower_case entirely from AutoTokenizer, as you suggested

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:72]: ill leave it as required. Our configs set the default, i think that is a good trade off, so its visible to user

File: nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py:36]: black's fault

Comment (titu1994) [reply nemo/collections/nlp/modules/common/huggingface/huggingface_utils.py:36]: Lol ok

File: nemo/collections/nlp/modules/common/tokenizer_utils.py
Comment (ekmb) [inline nemo/collections/nlp/modules/common/tokenizer_utils.py:53]: remove deleted args from docstring

File: nemo/collections/nlp/modules/common/tokenizer_utils.py
Comment (ekmb) [inline nemo/collections/nlp/modules/common/tokenizer_utils.py:78]: if not tokenizer_model or not os.path.exists(tokenizer_mode)..

File: nemo/collections/nlp/modules/common/tokenizer_utils.py
Comment (ekmb) [inline nemo/collections/nlp/modules/common/tokenizer_utils.py:87]: you can remove either this of the first check `f tokenizer_name not in get_tokenizer_list()`

**Review by ekmb — CHANGES_REQUESTED** [review]

**Comment by ekmb** [comment]
@yzhang123 could you please also change this Jenkins test to use SPE 'stage ('NER with BERT') '

File: nemo/collections/nlp/modules/common/tokenizer_utils.py
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/tokenizer_utils.py:87]: done

File: nemo/collections/nlp/modules/common/tokenizer_utils.py
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/tokenizer_utils.py:78]: thanks, I fixed this by removing this check here completely and moving `if not tokenizer_model or not os.path.exists(tokenizer_mode).` to SentencePieceTokenizer, so when user directly use SentencePieceTokenizer it also gets checked

File: nemo/collections/nlp/modules/common/tokenizer_utils.py
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/tokenizer_utils.py:53]: missed that, thanks!

File: nemo/collections/nlp/modules/common/megatron/megatron_utils.py
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/megatron/megatron_utils.py:181]: done, however FYI your others functions also do not have type

Comment (ekmb) [reply nemo/collections/nlp/modules/common/megatron/megatron_utils.py:181]: true, fixed already in my LM notebook PR

File: nemo/collections/nlp/modules/common/lm_utils.py
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/lm_utils.py:79]: since default_checkpoint_file was not used, after discussing will fix the bug in old code

File: nemo/collections/nlp/modules/common/megatron/megatron_utils.py
Comment (yzhang123) [reply nemo/collections/nlp/modules/common/megatron/megatron_utils.py:103]: done after discussing

File: nemo/collections/nlp/models/token_classification/token_classification_model.py
Comment (yzhang123) [reply nemo/collections/nlp/models/token_classification/token_classification_model.py:183]: done

**Comment by yzhang123** [comment]
> @yzhang123 could you please also change this Jenkins test to use SPE 'stage ('NER with BERT') '

I tested locally with your test and it works. Its not easy to add to ci, since I could not easily add special tokens for SPE from command line.  Please add a test at later point

**Comment by yzhang123** [comment]
> @yzhang123 could you please also change this Jenkins test to use SPE 'stage ('NER with BERT') '

**Comment by yzhang123** [comment]
> @yzhang123 could you please also change this Jenkins test to use SPE 'stage ('NER with BERT') '

tested offline and it works

**Review by titu1994 — APPROVED** [review]
LGTM

File: nemo/collections/common/tokenizers/tokenizer_spec.py
Comment (ekmb) [inline nemo/collections/common/tokenizers/tokenizer_spec.py:35]: please either revert or add an additional token_to_id method

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py
Comment (ekmb) [inline nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:217]: I didn't work with this SPE code, not sure if it's ok to delete these lines. 
Maybe leave for a separate PR and discuss with @AlexGrinch before this change

**Review by ekmb — CHANGES_REQUESTED** [review]

File: nemo/collections/common/tokenizers/sentencepiece_tokenizer.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/sentencepiece_tokenizer.py:217]: i removed the special tokens from create_spc_tokenizer, it works fine and it was never used for that anyway

File: nemo/collections/common/tokenizers/tokenizer_spec.py
Comment (yzhang123) [reply nemo/collections/common/tokenizers/tokenizer_spec.py]: done

**Review by ekmb — APPROVED** [review]

**Pull request merged**
